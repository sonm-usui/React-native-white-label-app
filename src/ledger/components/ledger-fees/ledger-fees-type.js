// @flow

import * as React from 'react'
import type { LedgerFees } from '../../type-ledger-store'
import type { ReduxConnect } from '../../../common/type-common'
import type { WalletBalance } from '../../../wallet/type-wallet'

type ReactComponent = React.Node
type NoOp = () => void

export type LedgerFeesProps = {
  transferAmount: string,
  ledgerFees: LedgerFees,
  walletBalance: WalletBalance,
  renderPhases?: LedgerFeesRenderPhases,
  render?: (
    state: LedgerFeesStateEnum,
    data?: LedgerFeesData,
    retry: NoOp
  ) => ReactComponent,
  onStateChange: (state: LedgerFeesStateEnum, data?: LedgerFeesData) => void,
} & ReduxConnect

export type LedgerFeesRenderPhases = {
  IN_PROGRESS: (data?: any, retry: NoOp) => ReactComponent,
  ERROR: (data?: any, retry: NoOp) => ReactComponent,
  ZERO_FEES: (data?: LedgerFeesData, retry: NoOp) => ReactComponent,
  TRANSFER_EQUAL_TO_BALANCE: (
    data?: LedgerFeesData,
    retry: NoOp
  ) => ReactComponent,
  TRANSFER_POSSIBLE_WITH_FEES: (
    data?: LedgerFeesData,
    retry: NoOp
  ) => ReactComponent,
  TRANSFER_NOT_POSSIBLE_WITH_FEES: (
    data?: LedgerFeesData,
    retry: NoOp
  ) => ReactComponent,
}

export type LedgerFeesStateEnum =
  | 'IN_PROGRESS'
  | 'ERROR'
  | 'ZERO_FEES'
  | 'TRANSFER_EQUAL_TO_BALANCE'
  | 'TRANSFER_POSSIBLE_WITH_FEES'
  | 'TRANSFER_NOT_POSSIBLE_WITH_FEES'

export type LedgerFeesData = {
  fees: string,
  total: string,
  currentTokenBalance: string,
}
