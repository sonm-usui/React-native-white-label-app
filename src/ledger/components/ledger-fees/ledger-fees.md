This component is created using render props concept. [See more details about render props](https://reactjs.org/docs/render-props.html). This component is created to share ledger fees related logic and not the UI. We can still create some UI components that can be exposed from same folder, but we think that this folder should only share logic for different states and rendering UI that belongs to those states. UI for different states should be passed by the consumer of this component.

### Usage:

```js
<LedgerFees
  renderPhases={{
    IN_PROGRESS: () => <Loader message="Fetching transaction fees..." />,
    ERROR: () => <Error message="" />,
    ZERO_FEES: (data?: LedgerFeesData, retry: () => void) => <ZeroFees />,
    TRANSFER_EQUAL_TO_BALANCE: (data?: LedgerFeesData, retry: () => void) => <TransferPossibleButNoBalanceLeft />,
    TRANSFER_POSSIBLE_WITH_FEES: (data?: LedgerFeesData, retry: () => void) => <TransferPossible />,
    TRANSFER_NOT_POSSIBLE_WITH_FEES: (data?: LedgerFeesData, retry: () => void) => <TransferNotPossible />,
  }}
  onStateChange={(state: LedgerFeesStatus, data?: LedgerFeesData) => console.log({state, data})}
  transferAmount={"0.0001"}
  render={(state: LedgerFeesStatus, data?: LedgerFeesData, retry: () => void) => <CommonComponentForAllStates />}
/>

// Types for data passed in renderPhase props
type LedgerFeesData = {
  fees: string,
  total: string,
  currentTokenBalance: string
}
// All of data that is passed in above type is number formatted
// if we need to make any calculations using above data
// then we need to use BigNumber constructor and pass above strings
// and then we can go ahead and perform those operations

// type for state that we pass in onStateChange fn passed by consumer
type LedgerFeesStatus = 'IN_PROGRESS' 
                        | 'ERROR'
                        | 'ZERO_FEES'
                        | 'TRANSFER_EQUAL_TO_BALANCE'
                        | 'TRANSFER_POSSIBLE_WITH_FEES'
                        | 'TRANSFER_NOT_POSSIBLE_WITH_FEES'
// Above is an Enum type that gets passed to fn passed by consumer
// it would one of the value, so that consumer can perform it's own operation
// on the basis of ledger fees state change
```
*Note*: For avoiding re-rendering, props that are passed in this component should not be created as new object on every render. In above example we are just creating new renderPhase object on every render. So, LedgerFees component will re-render on every re-render of parent component when it shouldn't because essentially components for each state are going to be the same. Same goes with `onStateChange` prop as well, this prop should also be a reference and not an arrow function inside render method to avoid unnecessary re-render.
