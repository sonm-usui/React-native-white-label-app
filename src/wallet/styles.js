// @flow
import { StyleSheet } from 'react-native'
import {
  PADDING_VERTICAL,
  isBiggerThanShortDevice,
  color,
} from '../common/styles/constant'

const ctaButtonHeight = isBiggerThanShortDevice ? 63 : 43

const styles = StyleSheet.create({
  headerCloseIcon: {
    marginTop: PADDING_VERTICAL,
  },
  ctaButton: {
    flex: -1,
    height: ctaButtonHeight,
    borderRadius: 5,
    backgroundColor: color.bg.eighth.color,
  },
  signTaaButton: {
    marginTop: 20,
    marginBottom: 6,
    marginHorizontal: '5%',
  },
  alignItemsCenter: {
    marginBottom: 6,
    marginHorizontal: '5%',
  },
  verticalSpacing: {
    flexDirection: 'column',
    alignContent: 'center',
    justifyContent: 'space-around',
  },
  horizontalSpacing: {
    flexDirection: 'column',
    alignContent: 'center',
    marginTop: 6,
    marginLeft: 6,
    marginRight: 6,
  },
})

export default styles
