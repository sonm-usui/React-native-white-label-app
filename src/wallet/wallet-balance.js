// @flow

import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { View } from 'react-native'
import type { WalletBalanceProps } from './type-wallet'
import type { Store } from '../store/type-store'
import { getWalletBalance } from '../store/store-selector'
import { refreshWalletBalance } from './wallet-store'

export class WalletBalance extends PureComponent<WalletBalanceProps, void> {
  componentDidMount() {
    this.props.refreshWalletBalance()
  }

  render() {
    const { walletBalance, render }: WalletBalanceProps = this.props
    return <View>{render(walletBalance)}</View>
  }
}

const mapStateToProps = (state: Store) => {
  return {
    walletBalance: getWalletBalance(state),
  }
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators({ refreshWalletBalance }, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(WalletBalance)
