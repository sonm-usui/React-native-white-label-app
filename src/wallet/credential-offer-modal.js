// @flow
import React, { PureComponent } from 'react'
import { StyleSheet, Platform } from 'react-native'
import { CustomView, Icon, CustomText, CustomButton } from '../components'
import { moderateScale } from 'react-native-size-matters'
import { colors } from '../common/styles/constant'
import Modal from 'react-native-modal'
import PaymentFailureModal from './payment-failure-modal'
import {
  CLAIM_REQUEST_STATUS,
  CREDENTIAL_OFFER_MODAL_STATUS,
} from '../claim-offer/type-claim-offer'
import type {
  CredentialOfferModalProps,
  CredentialOfferModalState,
} from '../claim-offer/type-claim-offer'
import { ClaimRequestStatusModal } from '../claim-offer/claim-request-modal'
import { LedgerFeesModal } from '../components/ledger-fees-modal/ledger-fees-modal'

export default class CredentialOfferModal extends PureComponent<
  CredentialOfferModalProps,
  CredentialOfferModalState
> {
  onContinue = () => {
    this.props.onClose()
  }

  render() {
    const {
      testID,
      claimRequestStatus,
      claimOfferData,
      isValid,
      logoUrl,
      payTokenValue,
      credentialOfferModalStatus,
      transferAmount,
      connectionName,
    } = this.props
    return (
      <Modal
        backdropColor={colors.gray5}
        isVisible={
          credentialOfferModalStatus !== CREDENTIAL_OFFER_MODAL_STATUS.NONE
        }
        animationIn="zoomIn"
        animationOut="zoomOut"
        animationOutTiming={100}
        useNativeDriver={true}
        hideModalContentWhileAnimating={true}
        onBackButtonPress={this.props.onClose}
        onBackdropPress={this.props.onClose}
        onModalHide={this.props.onModalHide}
        testID={'credential-offer-modal'}
        style={[
          claimRequestStatus ===
            CLAIM_REQUEST_STATUS.SENDING_PAID_CREDENTIAL_REQUEST ||
          claimRequestStatus === CLAIM_REQUEST_STATUS.SENDING_CLAIM_REQUEST ||
          claimRequestStatus === CLAIM_REQUEST_STATUS.SEND_CLAIM_REQUEST_SUCCESS
            ? {
                padding: 0,
                margin: 0,
              }
            : {},
        ]}
      >
        {isValid &&
          credentialOfferModalStatus ===
            CREDENTIAL_OFFER_MODAL_STATUS.INSUFFICIENT_BALANCE &&
          payTokenValue && (
            <CustomView center fifth shadow style={[styles.container]}>
              <CustomView
                spaceBetween
                style={[styles.innerContainer]}
                center
                verticalSpace
              >
                <CustomView center verticalSpace>
                  <Icon src={require('../images/alertInfo.png')} />
                </CustomView>
                <CustomText
                  transparentBg
                  style={[styles.fontBlack]}
                  center
                  demiBold
                >
                  You do not have enough tokens to purchase this credential
                </CustomText>
              </CustomView>
              <CustomButton
                fifth
                onPress={this.props.onClose}
                title={'Continue'}
                textStyle={{ fontWeight: 'bold', color: colors.main }}
              />
            </CustomView>
          )}

        {isValid &&
          credentialOfferModalStatus ===
            CREDENTIAL_OFFER_MODAL_STATUS.CREDENTIAL_REQUEST_FAIL && (
            <CustomView center fifth shadow style={[styles.container]}>
              <CustomView
                spaceBetween
                style={[styles.innerContainer]}
                center
                verticalSpace
              >
                <CustomView center verticalSpace>
                  <Icon src={require('../images/alertInfo.png')} />
                </CustomView>
                <CustomText
                  transparentBg
                  style={[styles.fontBlack]}
                  center
                  demiBold
                >
                  Error accepting credential. Please try again.
                </CustomText>
              </CustomView>
              <CustomButton
                fifth
                onPress={this.props.onClose}
                title={'OK'}
                textStyle={{ fontWeight: 'bold' }}
              />
            </CustomView>
          )}

        {claimRequestStatus &&
          claimOfferData &&
          isValid &&
          credentialOfferModalStatus ===
            CREDENTIAL_OFFER_MODAL_STATUS.CREDENTIAL_REQUEST_STATUS && (
            <ClaimRequestStatusModal
              claimRequestStatus={claimRequestStatus}
              payload={claimOfferData}
              onContinue={this.props.onClose}
              senderLogoUrl={logoUrl}
              payTokenValue={payTokenValue}
              message1={payTokenValue ? 'You paid' : 'You accepted'}
              message3={payTokenValue ? '' : 'from'}
              message6={
                payTokenValue ? 'They will issue it to you shortly.' : ''
              }
              isPending={
                claimRequestStatus ===
                  CLAIM_REQUEST_STATUS.SENDING_PAID_CREDENTIAL_REQUEST ||
                claimRequestStatus ===
                  CLAIM_REQUEST_STATUS.SENDING_CLAIM_REQUEST
              }
              onModalHide={this.props.onModalHide}
            />
          )}

        {credentialOfferModalStatus ===
          CREDENTIAL_OFFER_MODAL_STATUS.SEND_PAID_CREDENTIAL_REQUEST_FAIL ||
        credentialOfferModalStatus ===
          CREDENTIAL_OFFER_MODAL_STATUS.TOKEN_SENT_FAIL ? (
          <PaymentFailureModal
            connectionName={
              connectionName ||
              (claimOfferData != null && claimOfferData.issuer != null
                ? claimOfferData.issuer.name
                : '')
            }
            testID={`${testID}-payment-failure-modal`}
            onClose={this.props.onClose}
            onRetry={this.props.onRetry}
          />
        ) : null}

        {(payTokenValue || transferAmount) &&
        credentialOfferModalStatus ===
          CREDENTIAL_OFFER_MODAL_STATUS.LEDGER_FEES ? (
          <LedgerFeesModal
            onNo={this.props.onNo}
            onYes={this.props.onYes}
            transferAmount={transferAmount}
            renderFeesText={this.props.renderFeesText}
          />
        ) : null}
      </Modal>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginHorizontal: moderateScale(30),
  },
  innerContainer: {
    ...Platform.select({
      ios: {
        borderBottomColor: colors.gray1,
        borderBottomWidth: StyleSheet.hairlineWidth,
      },
      android: {
        borderBottomColor: colors.gray5,
        borderBottomWidth: 1,
      },
    }),
    padding: moderateScale(20),
  },
  message: {
    marginBottom: moderateScale(5),
  },
  fontBlack: {
    color: colors.gray2,
  },
})
