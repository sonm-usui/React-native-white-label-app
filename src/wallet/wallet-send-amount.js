// @flow

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { Animated } from 'react-native'
import { scale } from 'react-native-size-matters'
import { Container, CustomView, CustomText, CustomButton } from '../components'
import { Keyboard } from '../components'
import { color, MEDIUM_DEVICE, deviceHeight } from '../common/styles/constant'
import { SEND_TOKEN_BUTTON, FONT_SIZE_MAPPING } from './wallet-constants'
import styles from './styles'
import type {
  WalletSendAmountState,
  WalletSendAmountProps,
} from './type-wallet'
import { walletTabSendDetailsRoute } from '../common/route-constants'
import { selectTokenAmount } from './wallet-store'
import {
  getWalletBalance,
  getAlreadySignedAgreement,
  getThereIsANewAgreement,
} from '../store/store-selector'
import type { Store } from '../store/type-store'
import { STORE_STATUS } from './type-wallet'
import { txnAuthorAgreementRoute } from '../common/route-constants'
import { SEND_TAB, SEND_TAB_TEST_ID } from './wallet-constants'
import { conversionFactorLength } from '../sovrin-token/sovrin-token-converter'

class WalletSendAmount extends Component<
  WalletSendAmountProps,
  WalletSendAmountState
> {
  static navigationOptions = (navEvent: any) => ({
    tabBarLabel: SEND_TAB,
    tabBarTestIDProps: {
      testID: SEND_TAB_TEST_ID,
      accessible: true,
      accessibleLabel: SEND_TAB_TEST_ID,
      accessibilityLabel: SEND_TAB_TEST_ID,
    },
    tabBarOnPress: (onPressEvent) => {
      navEvent.route.params.onTabPress(onPressEvent, navEvent)
    },
  })

  _shake: any
  _tokenKeyboard: null | Keyboard
  state = {
    text: '',
  }

  constructor(props) {
    super(props)
    this._shake = new Animated.Value(0)
  }

  changeText = (text: string, animate: boolean) => {
    if (animate) {
      return this.shake()
    }
    return this.setState({ text })
  }

  shake = () => {
    this._shake.setValue(0)
    Animated.spring(this._shake, {
      toValue: 1,
      friction: 3,
      tension: 10,
      useNativeDriver: true,
    }).start(() => {
      this._shake.setValue(0)
    })
  }

  sendTokenAmount = () => {
    this.props.selectTokenAmount(this.state.text)
    if (this.state.text.length) {
      this.props.screenProps.navigation.navigate(walletTabSendDetailsRoute, {
        ...this.props.navigation,
      })
    }
  }

  saveTokenKeyboardRef = (tokenKeyboard) => {
    this._tokenKeyboard = tokenKeyboard
  }

  render() {
    const { text } = this.state

    // adjust fontSize based on text length
    const fontSize = scale(FONT_SIZE_MAPPING(text.length))
    const animatedStyle = {
      transform: [
        {
          translateX: this._shake.interpolate({
            inputRange: [0, 0.1, 0.2, 0.3, 0.4, 0.55, 0.7, 0.8, 0.9, 1],
            outputRange: [0, -10, 0, -10, 0, -10, 0, -10, 0, 0],
          }),
        },
      ],
    }
    const isSendDisabled = text.length < 1 || text === '0'

    if (!this.props.alreadySignedAgreement || this.props.thereIsANewAgreement) {
      return (
        <Container tertiary style={[styles.horizontalSpacing]}>
          <CustomView verticalSpace>
            <CustomText h6 tertiary demibold transparentBg>
              Please first accept the Transaction Author Agreement (TAA)!
            </CustomText>
            <CustomView style={[styles.signTaaButton]}>
              <CustomButton
                disabled={false}
                customColor={{ backgroundColor: color.bg.eighth.color }}
                onPress={this.handleTabPress}
                testID="show-txn-author-agreement"
                style={[styles.ctaButton]}
                primary
                title="Read and Sign TAA"
              />
            </CustomView>
          </CustomView>
        </Container>
      )
    } else {
      return (
        <Container tertiary style={[styles.verticalSpacing]}>
          <CustomView>
            <CustomView verticalSpace>
              <CustomText
                animated
                formatNumber
                transparentBg
                center
                style={[
                  animatedStyle,
                  {
                    fontSize,
                    color: color.bg.seventh.font.fifth,
                    height: deviceHeight < MEDIUM_DEVICE ? 70 : 90,
                    lineHeight: deviceHeight < MEDIUM_DEVICE ? 75 : 80,
                  },
                ]}
                numberOfLines={1}
              >
                {text || 0}
              </CustomText>
            </CustomView>
            <CustomText h6 center tertiary demibold uppercase transparentBg>
              sovrin tokens
            </CustomText>
          </CustomView>
          <Keyboard
            color={color.bg.seventh.font.fifth}
            onPress={this.changeText}
            maxValue={this.props.walletBalance}
            ref={this.saveTokenKeyboardRef}
            afterDecimalSeparatorMaxLength={conversionFactorLength}
          />
          <CustomView safeArea style={[styles.alignItemsCenter]}>
            <CustomButton
              disabled={isSendDisabled}
              customColor={{ backgroundColor: color.bg.eighth.color }}
              onPress={this.sendTokenAmount}
              testID={SEND_TOKEN_BUTTON}
              style={[styles.ctaButton]}
              primary
              title="Select Recipient"
            />
          </CustomView>
        </Container>
      )
    }
  }

  componentDidUpdate(prevProps: WalletSendAmountProps) {
    if (
      this.props.paymentStatus !== prevProps.paymentStatus &&
      this.props.paymentStatus === STORE_STATUS.SUCCESS
    ) {
      // if payment is successful, we can now reset token amount and keyboard
      this.setState({ text: '' })
      this._tokenKeyboard && this._tokenKeyboard.clear()
    }
  }

  componentDidMount() {
    this.props.navigation.setParams({
      onTabPress: this.handleTabPress,
    })
  }

  handleTabPress = () => {
    if (!this.props.alreadySignedAgreement || this.props.thereIsANewAgreement) {
      this.props.screenProps.navigation.navigate(txnAuthorAgreementRoute)
    }
  }
}

const mapStateToProps = (state: Store) => {
  return {
    thereIsANewAgreement: getThereIsANewAgreement(state),
    alreadySignedAgreement: getAlreadySignedAgreement(state),
    walletBalance: getWalletBalance(state),
    paymentStatus: state.wallet.payment.status,
  }
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      selectTokenAmount,
    },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(WalletSendAmount)
