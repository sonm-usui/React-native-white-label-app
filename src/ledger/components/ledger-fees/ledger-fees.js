// @flow

import { Component } from 'react'
import { connect } from 'react-redux'
import { BigNumber } from 'bignumber.js'

import type { Store } from '../../../store/type-store'
import type {
  LedgerFeesProps,
  LedgerFeesStateEnum,
  LedgerFeesData,
} from './ledger-fees-type'

import { getLedgerFees } from '../../type-ledger-store'
import { refreshWalletBalance } from '../../../wallet/wallet-store'
import { STORE_STATUS } from '../../../wallet/type-wallet'

class LedgerFeesComponent extends Component<LedgerFeesProps, void> {
  render() {
    const { ledgerFees, transferAmount, walletBalance } = this.props
    const { state, data } = getLedgerFeeState(
      ledgerFees,
      walletBalance,
      transferAmount
    )

    if (this.props.render) {
      return this.props.render(state, data, this.retry)
    }

    if (this.props.renderPhases) {
      return this.props.renderPhases[state](data, this.retry)
    }

    return null
  }

  componentDidUpdate(prevProps) {
    const { ledgerFees, transferAmount, walletBalance } = this.props

    if (
      prevProps.ledgerFees !== ledgerFees ||
      prevProps.walletBalance !== walletBalance ||
      transferAmount !== prevProps.transferAmount
    ) {
      // whenever state of ledgerFees or walletBalance changes
      // or transfer amount changes
      // we need to update state and should notify consumer of this component
      this.updateConsumer()
    }
  }

  componentDidMount() {
    // our component always starts in progress state
    // so we update consumer of this component to be in progress state
    this.props.onStateChange('IN_PROGRESS')
  }

  UNSAFE_componentWillMount() {
    // We agree that we should not use UNSAFE_componentWillMount lifecycle
    // events of react component, but there are good reasons for using
    // this event here.
    // When we render this component, we render UI on the basis of ledger fees
    // and wallet balance from redux store. But, data in redux store could be
    // stale data, and it might not have been refreshed or even hydrated
    // Example of these cases are if, user goes to paid cred screen directly
    // and wallet balance is only refreshed if user goes to settings screen
    // or to token screen. So, we can see that how user can get stale data
    // if we load this component in any of screen that does not involve wallet
    // balance UI flow.
    // Now, we know that data can be stale while we run `render` method of this
    // component. But this still does not explain why we are using UNSAFE_*
    // lifecycle event.
    // If data is stale, then inside render method we could show wrong fees
    // to user and it will be incorrect and has unintended effect of user
    // raising bug about incorrect balance deduction.
    // So, now we know that we could show wrong fees and some possible bugs.
    // Let us discuss the some possible ways we can fix this issue
    // 1. We can ensure that ledgerFees and walletBalance are updated
    //    before we render this component
    // 2. We can start fetching ledgerFees and walletBalance in
    //    componentDidMount() { this.props.getLedgerFees() }
    // 3. We can initialize a state variable and use it in render method
    //    to decide if we have updated ledgerFees at least once or not. e.g.
    //    constructor() { this.state = { refreshStarted: false } }
    //    render() { if (this.state.refreshStarted) {  } else { } }
    //    componentDidMount() { this.getLedgerFees(); this.setState({re:true})}
    // All of above approaches has one or other issue.
    // 1. It is difficult to ensure, because developers won't remember this
    //    critical piece before using this component. User of this component
    //    has too much responsibility. So, we can't use 1st approach
    // 2. componentDidMount is run after render method. So app user might still
    //    see a flash of UI change which is bad UX. So, we can't use 2nd either
    // 3. This approach comes pretty close to meeting all of our requirements
    //    of refreshing data and also showing correct UI with correct UX. But,
    //    the problem with this approach is that this component is based on
    //    render props implementation. So, if we set state, then component will
    //    re-render on state change, so our component will re-render 3 times
    //    one initial render
    //    second, getLedgerFees changes redux store, which triggers re-render
    //    third, state change to specify that we have requested refresh
    //    So, although 3rd approach meets our requirement but does it at a cost
    // Now, we come to the point on why we are using UNSAFE_* method
    // UNSAFE_componentWillMount, is called before render method AND any
    // state change or prop change does not trigger re-render due to the way
    // UNSAFE_componentWillMount behaves. Now, we can see how this solves all
    // our problems that we were describing. We can ask ourselves that if this
    // lifecycle event is so great, why is this UNSAFE, and why people don't
    // use it often? The reason is that since React re-wrote their algorithm
    // to use Fiber, componentWillMount can be called multiple times.
    // And if we are triggering side effects or subscribing to something
    // we might end up subscribing and sending API calls multiple times
    // that is why this event is marked as UNSAFE.
    // Now, we come to the question that how are we going to ensure that we are
    // not hit by UNSAFE issues. In our componentWillMount, we are asking for
    // refreshing balance and ledger fees. So, our ledger store is already
    // taking care of this, imagine that any logic or component
    // can ask to refresh ledger fees, so this is not a problem that we need to
    // handle from a component, this problem lies with our redux-saga
    // Fortunately, in redux-saga we have `takeLeading` that fixes this problem
    // elegantly. If we use `takeLeading`, then even if we ask for
    // refresh 10 times, redux-saga will ensure that till ir finishes refresh
    // no other request are not entertained. So, we are safe by redux-saga
    // and hence we can use this lifecycle method safely for us.
    this.props.dispatch(getLedgerFees())
    this.props.dispatch(refreshWalletBalance())
  }

  updateConsumer() {
    const { ledgerFees, transferAmount, walletBalance } = this.props
    const { state, data } = getLedgerFeeState(
      ledgerFees,
      walletBalance,
      transferAmount
    )
    this.props.onStateChange(state, data)
  }

  retry = () => {
    this.props.dispatch(getLedgerFees())
  }
}

const zeroAmount = new BigNumber('0')
function getLedgerFeeState(
  ledgerFees: $PropertyType<LedgerFeesProps, 'ledgerFees'>,
  walletBalance: $PropertyType<LedgerFeesProps, 'walletBalance'>,
  toTransfer: $PropertyType<LedgerFeesProps, 'transferAmount'>
): { state: LedgerFeesStateEnum, data?: LedgerFeesData } {
  const {
    data: { transfer },
    status: feesStatus,
  } = ledgerFees
  const { data: balance, status: balanceStatus } = walletBalance

  switch (true) {
    case [feesStatus, balanceStatus].includes(STORE_STATUS.IDLE):
    case [feesStatus, balanceStatus].includes(STORE_STATUS.IN_PROGRESS):
      return { state: 'IN_PROGRESS' }

    case feesStatus === STORE_STATUS.SUCCESS &&
      balanceStatus === STORE_STATUS.SUCCESS: {
      const feesAmount = new BigNumber(transfer)
      if (feesAmount.isLessThanOrEqualTo(zeroAmount)) {
        return { state: 'ZERO_FEES' }
      }

      const walletBalanceAmount = new BigNumber(balance)
      const transferAmount = new BigNumber(toTransfer)
      const totalAmountNeeded = transferAmount.plus(feesAmount)
      const feesData = {
        fees: formatToken(feesAmount),
        total: formatToken(totalAmountNeeded),
        currentTokenBalance: formatToken(walletBalanceAmount),
      }

      if (walletBalanceAmount.isLessThan(totalAmountNeeded)) {
        // if wallet does not enough balance to pay for transfer amount and fees
        // then we need to tell user that this transaction cannot be done
        return { state: 'TRANSFER_NOT_POSSIBLE_WITH_FEES', data: feesData }
      }

      if (walletBalanceAmount.isEqualTo(totalAmountNeeded)) {
        return { state: 'TRANSFER_EQUAL_TO_BALANCE', data: feesData }
      }

      return { state: 'TRANSFER_POSSIBLE_WITH_FEES', data: feesData }
    }

    case [feesStatus, balanceStatus].includes(STORE_STATUS.ERROR):

    default:
      return { state: 'ERROR' }
  }
}

function formatToken(amount: BigNumber): string {
  return amount.toFixed(8)
}

const mapStateToProps = (state: Store) => ({
  ledgerFees: state.ledger.fees,
  walletBalance: state.wallet.walletBalance,
})

export const LedgerFees = connect(mapStateToProps)(LedgerFeesComponent)
