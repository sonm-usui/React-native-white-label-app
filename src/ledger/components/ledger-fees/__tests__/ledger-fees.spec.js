// @flow

import React from 'react'
import { View, Text } from 'react-native'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import { LedgerFees } from '../ledger-fees'
import {
  getStore,
  walletBalance as mockWalletBalance,
} from '../../../../../__mocks__/static-data'
import { ledgerStoreWithTransferFees as mockLedgerWithFees } from '../../../../../__mocks__/data/ledger-store-mock-data'
import { STORE_STATUS } from '../../../../wallet/type-wallet'
import { getLedgerFees } from '../../../type-ledger-store'
import { refreshWalletBalance } from '../../../../wallet/wallet-store'

describe('<LedgerFees />', () => {
  it('should refresh wallet balance and ledger fees', () => {
    const { store } = setup()
    expect(store.dispatch).toHaveBeenCalledWith(getLedgerFees())
    expect(store.dispatch).toHaveBeenCalledWith(refreshWalletBalance())
  })

  it('should render IN_PROGRESS on initial mount', () => {
    const { props } = setup()
    expect(props.onStateChange).toHaveBeenCalledWith('IN_PROGRESS')
  })

  it('should render IN_PROGRESS component if balance is in progress', () => {
    const { wrapper } = setup({ balance: { status: STORE_STATUS.IN_PROGRESS } })
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should render IN_PROGRESS component if fees is in progress', () => {
    const { wrapper } = setup({ fees: { status: STORE_STATUS.IN_PROGRESS } })
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should render IN_PROGRESS component if both balance and fees are in progress', () => {
    const { wrapper } = setup({
      balance: { status: STORE_STATUS.IN_PROGRESS },
      fees: { status: STORE_STATUS.IN_PROGRESS },
    })
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should render ZERO component if fees is zero', () => {
    const { wrapper } = setup({
      fees: { data: { transfer: '0' } },
    })
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should call TRANSFER_POSSIBLE_WITH_FEES with correct fees data', () => {
    const { wrapper } = setup()
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should call TRANSFER_NOT_POSSIBLE_WITH_FEES with correct fees data', () => {
    const { wrapper } = setup({
      fees: { data: { transfer: '10000' } },
    })
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  // TODO:KS Add tests which check functionality for render method as well

  function getText(text: string) {
    return (
      <View>
        <Text>{text}</Text>
      </View>
    )
  }

  function getProps() {
    return {
      transferAmount: '0.001',
      renderPhases: {
        IN_PROGRESS: jest.fn(() => getText('In progress...')),
        ERROR: jest.fn(() => getText('error')),
        ZERO_FEES: jest.fn(() => getText('zero fees')),
        TRANSFER_EQUAL_TO_BALANCE: jest.fn(() =>
          getText('TRANSFER_EQUAL_TO_BALANCE')
        ),
        TRANSFER_POSSIBLE_WITH_FEES: jest.fn(() =>
          getText('TRANSFER_POSSIBLE_WITH_FEES')
        ),
        TRANSFER_NOT_POSSIBLE_WITH_FEES: jest.fn(() =>
          getText('TRANSFER_NOT_POSSIBLE_WITH_FEES')
        ),
      },
      onStateChange: jest.fn(),
    }
  }

  function setup(extraProps?: { balance?: Object, fees?: Object } = {}) {
    const store = getStore({
      wallet: {
        walletBalance: {
          ...mockWalletBalance,
          ...(extraProps.balance || {}),
        },
      },
      ledger: {
        fees: {
          ...mockLedgerWithFees.fees,
          ...(extraProps.fees || {}),
        },
      },
    })

    const props = getProps()
    const wrapper = renderer.create(
      <Provider store={store}>
        <LedgerFees {...props} />
      </Provider>
    )

    return { props, wrapper, store }
  }
})
