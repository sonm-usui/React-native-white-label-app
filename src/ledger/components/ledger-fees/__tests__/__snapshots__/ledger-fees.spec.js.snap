// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LedgerFees /> should call TRANSFER_NOT_POSSIBLE_WITH_FEES with correct fees data 1`] = `
<View>
  <Text>
    TRANSFER_NOT_POSSIBLE_WITH_FEES
  </Text>
</View>
`;

exports[`<LedgerFees /> should call TRANSFER_POSSIBLE_WITH_FEES with correct fees data 1`] = `
<View>
  <Text>
    TRANSFER_POSSIBLE_WITH_FEES
  </Text>
</View>
`;

exports[`<LedgerFees /> should render IN_PROGRESS component if balance is in progress 1`] = `
<View>
  <Text>
    In progress...
  </Text>
</View>
`;

exports[`<LedgerFees /> should render IN_PROGRESS component if both balance and fees are in progress 1`] = `
<View>
  <Text>
    In progress...
  </Text>
</View>
`;

exports[`<LedgerFees /> should render IN_PROGRESS component if fees is in progress 1`] = `
<View>
  <Text>
    In progress...
  </Text>
</View>
`;

exports[`<LedgerFees /> should render ZERO component if fees is zero 1`] = `
<View>
  <Text>
    zero fees
  </Text>
</View>
`;
